File: iOSSDK/SDK+Unity/BTBaseSDK+UnityMessageReceiver.m
//
//  BTBaseSDK+UnityMessageReceiver.m
//  BTBaseSDK
//
//  Created by Alex Chow on 2018/6/16.
//  Copyright © 2018年 btbase. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "BTBaseSDK+UnityMessageReceiver.h"

NSNotificationName const UnityMessageReceiverNotificationName = @"UnityMessageReceiverNotification";
NSString* kUnityMessageReceiverMessageName = @"kUnityMessageReceiverMessageName";
NSString* kUnityMessageReceiverMessageUserInfo = @"kUnityMessageReceiverMessageUserInfo";

@implementation BTBaseSDK(UnityMessageReceiver)
NSMutableArray* observers;
+(void)sendMessageToUnityMessageReceiver:(NSString *)msgName userInfo:(NSDictionary *)uinfo{
    NSMutableDictionary* info = [NSMutableDictionary alloc];
    if (uinfo) {
        [info setObject:uinfo forKey:kUnityMessageReceiverMessageUserInfo];
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:UnityMessageReceiverNotificationName object:nil userInfo:info];
}

+(void)startListenBTBaseNotifications{
    if (observers) {
        [self stopListenBTBaseNotifications];
    }else{
        observers = [NSMutableArray alloc];
    }
    NSObject* observer;
    observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"BTSessionService_onSessionInvalid" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        //TODO:
    }];
    [observers addObject:observer];
    
    observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"BTSessionService_onSessionInvalid" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        //TODO:
    }];
    [observers addObject:observer];
    
    observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"BTBaseHomeEntryDidShown" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        //TODO:
    }];
    [observers addObject:observer];
    
    observer = [[NSNotificationCenter defaultCenter] addObserverForName:@"BTBaseHomeEntryClosed" object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        //TODO:
    }];
    [observers addObject:observer];
}

+(void)stopListenBTBaseNotifications{
    for (NSObject* observer in observers) {
        [[NSNotificationCenter defaultCenter ] removeObserver:observer];
    }
}
@end

File: iOSSDK/SDK+Unity/BTBaseSDK+UnityMessageReceiver.swift
//
//  BTBaseSDK+UnityMessageReceiver.swift
//  BTBaseSDK
//
//  Created by Alex Chow on 2018/6/13.
//  Copyright © 2018年 btbase. All rights reserved.
//

import Foundation

// MARK: Message To UnityMessageReceiver

let UnityMessageReceiverNotificationName = Notification.Name("UnityMessageReceiverNotification")
let kUnityMessageReceiverMessageName = "kUnityMessageReceiverMessageName"
let kUnityMessageReceiverMessageUserInfo = "kUnityMessageReceiverMessageUserInfo"

public extension BTBaseSDK {
    @objc public static func sendMessageToUnityMessageReceiver(_ msgName: String, _ userInfo: NSDictionary?) {
        var info: [AnyHashable: Any] = [kUnityMessageReceiverMessageName: msgName]
        if let ui = userInfo {
            info[kUnityMessageReceiverMessageUserInfo] = ui
        }
        NotificationCenter.default.post(name: UnityMessageReceiverNotificationName, object: nil, userInfo: info)
    }
}

private var observers = [NSObjectProtocol]()
public extension BTBaseSDK {
    @objc public static func startListenBTBaseNotifications() {
        observers.append(NotificationCenter.default.addObserver(forName: BTSessionService.onSessionInvalid, object: nil, queue: nil, using: { (a) in
            //TODO:
        }))
        
        observers.append(NotificationCenter.default.addObserver(forName: BTSessionService.onSessionUpdated, object: nil, queue: nil, using: { (a) in
            //TODO:
        }))
        
        observers.append(NotificationCenter.default.addObserver(forName: Notification.Name.BTBaseHomeEntryDidShown, object: nil, queue: nil, using: { (a) in
            //TODO:
        }))
        
        observers.append(NotificationCenter.default.addObserver(forName: Notification.Name.BTBaseHomeEntryClosed, object: nil, queue: nil, using: { (a) in
            //TODO:
        }))
    }

    @objc public static func stopListenBTBaseNotifications() {
        for ob in observers {
            NotificationCenter.default.removeObserver(ob)
        }
    }
}
